### PR TITLE
Filter out voided decision issues

### LINKED ISSUE #11641: EP930 - Void decision issues for corrected request issues
### Notes:
* Decision issues and request issues have a many-to-many relationship
* Void a decision issue only if it is not connected to another request issue
* Add a status field to a decision issue: `closed_status`
* Set `voided` closed status when a decision issue is voided
* We will reuse this status for issues that are soft deleted as well 

### Open questions:
* Should we also void a request_decision_issue?

### Diagram

image

### AC:
Void decision issues for corrected request issues

### LINKED ISSUE #11642: EP930 - Filter out voided ratings and decision issues from contestable issues
### Notes: 
Methods: 
* `contestable_issues_from_decision_issues`
Filter out voided decision issues, or update decision_issue.finalized? to not include voided decision issues.
* `contestable_issues_from_ratings`
 Filter out rating issues where the rating_issue_reference_id matches any of the voided decision issue rating_issue_reference_ids

### AC: 
Filter out voided ratings and decision issues from contestable issues

### PR DESCRIPTION
**Pull request by aroltsch**
Resolves #11641 
Resolves #11642 

### Description
Filter out voided decision issues from contestable issues

### CODE REVIEW

File: app/models/decision_review.rb [hunk 1]
```diff
@@ -198,7 +198,7 @@ def establish!
   def contestable_issues
     return contestable_issues_from_decision_issues unless can_contest_rating_issues?
 
-    contestable_issues_from_ratings + contestable_issues_from_decision_issues
+    contestable_issues_from_ratings + contestable_issues_from_decision_issues.reject(&:voided?)
```
Comment (aroltsch) [inline app/models/decision_review.rb]: Regarding this [ticket](https://github.com/department-of-veterans-affairs/caseflow/issues/11642), we already filter out rating issues when the rating_issue_reference_id matches any of the decision issue rating_issue_reference_ids so we just need to reject decision issues that are voided here at the end.... I think..

File: app/workflows/contestable_issue_generator.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    filtered_from_ratings + filtered_from_decision_issues
+  end
+
+  private
+
+  attr_reader :review, :participant_id
+
+  def filtered_from_ratings
+    unfiltered_from_ratings.reject do |contestable_issue|
+      from_decision_issues.any? do |potential_duplicate|
```
Comment (leikkisa) [inline app/workflows/contestable_issue_generator.rb]: This looks good! This was the main thing I was watching, and you filter out the dupes before excluding the voided decisions.

File: app/workflows/contestable_issue_generator.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  def initialize(review, participant_id)
+    @review = review
+    @participant_id = participant_id
+  end
+
+  def contestable_issues
+    return [] unless review.receipt_date
+    return filtered_from_decision_issues unless review.can_contest_rating_issues?
+
+    filtered_from_ratings + filtered_from_decision_issues
```
Comment (leikkisa) [inline app/workflows/contestable_issue_generator.rb]: I know this is an artifact from before, but I think the naming of these is not very intuitive, so I'll make some (optional) suggestions.

With naming, I think it'd be nice to call these "contestable_ratings" and "contestable_decision_issues".

File: app/workflows/contestable_issue_generator.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  private
+
+  attr_reader :review, :participant_id
+
+  def filtered_from_ratings
```
Comment (leikkisa) [inline app/workflows/contestable_issue_generator.rb]: How about ratings_without_decision_issues

File: app/workflows/contestable_issue_generator.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    unfiltered_from_ratings.reject do |contestable_issue|
+      from_decision_issues.any? do |potential_duplicate|
+        contestable_issue.rating_issue_reference_id == potential_duplicate.rating_issue_reference_id
+      end
+    end
+  end
+
+  def filtered_from_decision_issues
```
Comment (leikkisa) [inline app/workflows/contestable_issue_generator.rb]: how about valid_decision_issues

File: app/workflows/contestable_issue_generator.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    from_decision_issues.reject { |contestable_issue| contestable_issue.decision_issue&.voided? }
+  end
+
+  def unfiltered_from_ratings
```
Comment (leikkisa) [inline app/workflows/contestable_issue_generator.rb]: Maybe eligible_ratings ?
ratings_before_receipt_date?

File: app/workflows/contestable_issue_generator.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      .map { |rating_issue| ContestableIssue.from_rating_issue(rating_issue, review) }
+  end
+
+  def from_decision_issues
+    @from_decision_issues ||= finalized_decision_issues.map do |decision_issue|
+      ContestableIssue.from_decision_issue(decision_issue, review)
+    end
+  end
+
+  def finalized_decision_issues
```
Comment (leikkisa) [inline app/workflows/contestable_issue_generator.rb]: eligible_finalized_decision_issues?

File: app/workflows/contestable_issue_generator.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      .map { |rating_issue| ContestableIssue.from_rating_issue(rating_issue, review) }
+  end
+
+  def from_decision_issues
+    @from_decision_issues ||= finalized_decision_issues_before_receipt_date.map do |decision_issue|
+      ContestableIssue.from_decision_issue(decision_issue, review)
+    end
+  end
+
+  def rating_issues_for_review
```
Comment (leikkisa) [inline app/workflows/contestable_issue_generator.rb]: This could probably be rating_issues, since they're not specific to this review.

**Pull request merged**